### ISSUE TITLE
How to install and test?

### ISSUE THREAD

**Issue by WolfgangFahl** [samwilson/diagrams-service#4]
When trying to use it i had no clue how to install and test this. Could you please elaborate a bit. The unit tests look good and I'd love to use and extend them.

The instruction "Set up your webserver to serve the public/ directory" looks insufficient since the PHP code supplied by composer in the vendor directory will be missing. What's the simplest way to overcome that?

**Comment by samwilson** [comment]
Good point. The docs were written rather hastily! I've made a couple of edits just now, and will make more.

The installation is the same as for most modern PHP applications. The line about serving the `public/` directory was confusing I think, because I didn't mean to imply that *only* that directory should exist on the server. And in fact, the app works fine even with the [docroot](https://httpd.apache.org/docs/current/mod/core.html#documentroot) at the root (or above) of the repository; it's only an extra security measure and for shorter URLs that I suggest that `public/` be the docroot.

**Comment by WolfgangFahl** [comment]
Needed 
`sudo apt install composer
`
then
` composer install --no-dev -o
`
failed 
```
 Problem 1
    - Installation request for wikimedia/toolforge-bundle dev-symfony5 -> satisfiable by wikimedia/toolforge-bundle[dev-symfony5].
    - wikimedia/toolforge-bundle dev-symfony5 requires ext-intl * -> the requested PHP extension intl is missing from your system.

  To enable extensions, verify that they are enabled in your .ini files:
    - /etc/php/7.2/cli/php.ini
    - /etc/php/7.2/cli/conf.d/10-mysqlnd.ini
    - /etc/php/7.2/cli/conf.d/10-opcache.ini
    - /etc/php/7.2/cli/conf.d/10-pdo.ini
    - /etc/php/7.2/cli/conf.d/15-xml.ini
    - /etc/php/7.2/cli/conf.d/20-calendar.ini
    - /etc/php/7.2/cli/conf.d/20-ctype.ini
    - /etc/php/7.2/cli/conf.d/20-curl.ini
    - /etc/php/7.2/cli/conf.d/20-dom.ini
    - /etc/php/7.2/cli/conf.d/20-exif.ini
    - /etc/php/7.2/cli/conf.d/20-fileinfo.ini
    - /etc/php/7.2/cli/conf.d/20-ftp.ini
    - /etc/php/7.2/cli/conf.d/20-gettext.ini
    - /etc/php/7.2/cli/conf.d/20-iconv.ini
    - /etc/php/7.2/cli/conf.d/20-json.ini
    - /etc/php/7.2/cli/conf.d/20-mbstring.ini
    - /etc/php/7.2/cli/conf.d/20-mysqli.ini
    - /etc/php/7.2/cli/conf.d/20-pdo_mysql.ini
    - /etc/php/7.2/cli/conf.d/20-phar.ini
    - /etc/php/7.2/cli/conf.d/20-posix.ini
    - /etc/php/7.2/cli/conf.d/20-readline.ini
    - /etc/php/7.2/cli/conf.d/20-shmop.ini
    - /etc/php/7.2/cli/conf.d/20-simplexml.ini
    - /etc/php/7.2/cli/conf.d/20-sockets.ini
    - /etc/php/7.2/cli/conf.d/20-sysvmsg.ini
    - /etc/php/7.2/cli/conf.d/20-sysvsem.ini
    - /etc/php/7.2/cli/conf.d/20-sysvshm.ini
    - /etc/php/7.2/cli/conf.d/20-tokenizer.ini
    - /etc/php/7.2/cli/conf.d/20-wddx.ini
    - /etc/php/7.2/cli/conf.d/20-xdebug.ini
    - /etc/php/7.2/cli/conf.d/20-xmlreader.ini
    - /etc/php/7.2/cli/conf.d/20-xmlwriter.ini
    - /etc/php/7.2/cli/conf.d/20-xsl.ini
  You can also run `php --ini` inside terminal to see which files are used by PHP in CLI mode.
```

**Comment by WolfgangFahl** [comment]
workaround

`sudo apt-get install php-intl
`

**Comment by WolfgangFahl** [comment]
When the public directory is served alone you get a 500 error code. As it looks 

- public
- vendor
- template

might be needed but i didn't test that yet.

**Comment by WolfgangFahl** [comment]
How can the tests be started from the command line or an IDE? What IDE do you recommend for debugging? Please document all these details.

**Comment by WolfgangFahl** [comment]
I still get

```
Oops! An Error Occurred
The server returned a "500 Internal Server Error".

Something is broken. Please let us know what you were doing when this error occurred. We will fix it as soon as possible. Sorry for any inconvenience caused.
```
when serving the whole directory ... the 500 error is not helpful at all. Is there a debug / log option to find out what's going on? The apache log also only has 1 line with a 500 error code.

**Comment by WolfgangFahl** [comment]
My recent experience is "modern" -> "not working" more often than not. More complexity, more components and higher speed of changes of parts unfortunately also means higher risk of failure. Thats why the unit and integration tests might be so helpful if i'd be able to run them myself.

**Comment by WolfgangFahl** [comment]
I tried:

```
sudo apt install node-grunt-cli
grunt test
```
and got:
```
grunt-cli: The grunt command line interface (v1.2.0)

Fatal error: Unable to find local grunt.

```

There seems to be a package.json for the CI. This project uses quite a bit of infrastructure and it would be good to get an overview on how and why this infrastructure is in place.

**Comment by samwilson** [comment]
Do you have Node and NPM installed? If so, try:

```
npm install
npm test
```

Sorry that I haven't had time in the last few days to work on the documentation here; I do intend to, soon.

**Comment by WolfgangFahl** [comment]
* https://github.com/gruntjs/grunt-docs/issues/38
* https://askubuntu.com/questions/1088662/npm-depends-node-gyp-0-10-9-but-it-is-not-going-to-be-installed
```
sudo apt install npm
sudo apt-get install nodejs-dev node-gyp libssl1.0-dev
npm install grunt --save-dev
```
```
grunt test
>> Local Npm module "grunt-banana-checker" not found. Is it installed?
>> Local Npm module "grunt-eslint" not found. Is it installed?
>> Local Npm module "grunt-stylelint" not found. Is it installed?
Warning: Task "eslint" not found. Use --force to continue.

Aborted due to warnings.
```

**Comment by samwilson** [comment]
Why are you only installing grunt? You should just run `npm install` not `npm install grunt --save-dev`.

**Comment by WolfgangFahl** [comment]
Our comments overlapped - having no clue what to do I looked at the package.json and saw the "grunt test" in there assuming that this would be a valid way to test. 

After having overcome all the dependency and broken component obstacles on the virtual machine I am now at:
```
npm test

> @ test /home/wf/source/php/diagrams-service
> grunt test

Running "eslint:all" (eslint) task

Running "stylelint:all" (stylelint) task
>> Linted 1 files without errors

Running "banana:all" (banana) task
>> 1 message directory checked.

Done.
```

But still the 500 error persists. What needs to be done to actually have the service working?

**Comment by samwilson** [comment]
Do you get any errors when you run `composer install`?

**Comment by WolfgangFahl** [comment]
```
composer install
Cannot create cache directory /home/wf/.composer/cache/repo/https---packagist.org/, or directory is not writable. Proceeding without cache
Cannot create cache directory /home/wf/.composer/cache/files/, or directory is not writable. Proceeding without cache
Cannot create cache directory /home/wf/.composer/cache/repo/https---flex.symfony.com/, or directory is not writable. Proceeding without cache
Cannot create cache directory /home/wf/.composer/cache/repo/https---packagist.org/, or directory is not writable. Proceeding without cache
Cannot create cache directory /home/wf/.composer/cache/repo/https---packagist.org/, or directory is not writable. Proceeding without cache
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file

Prefetching 3 packages 🎶
  - Downloading (100%)

Package operations: 8 installs, 0 updates, 0 removals
  - Installing squizlabs/php_codesniffer (3.5.0): Downloading (100%)         
  - Installing composer/spdx-licenses (1.5.1): Downloading (100%)         
  - Installing composer/semver (1.5.0): Downloading (100%)         
  - Installing mediawiki/mediawiki-codesniffer (v28.0.0): Downloading (100%)      - Installing symfony/dom-crawler (v5.0.2): Downloading (100%)         
  - Installing symfony/browser-kit (v5.0.2): Downloading (100%)         
  - Installing symfony/css-selector (v5.0.2): Downloading (100%)         
  - Installing symfony/phpunit-bridge (v5.0.2): Downloading (100%)         
Generating autoload files
Executing script cache:clear [OK]
```

**Comment by WolfgangFahl** [comment]
```
Oops! An Error Occurred
The server returned a "500 Internal Server Error".

Something is broken. Please let us know what you were doing when this error occurred. We will fix it as soon as possible. Sorry for any inconvenience caused.
```

**Comment by samwilson** [comment]
And you have nothing in your logs about what the actual error is? How about when you set `APP_ENV=dev` in `.env`?

**Comment by samwilson** [comment]
It sounds like it might be a permissions issue. Does the web server user have write permissions to the var directory?

**Comment by WolfgangFahl** [comment]
Which log are you talking about - the apache error log / system log / php log? 
What part of var are you talking about?
The vm is serving the mediawiki and eclipse debug environment for http://wiki.bitplan.com/index.php/S5SlideShow just fine. I was hoping to get the GraphViz extension debugged in the same enviroment but with the separate service its a nightmare for me

**Comment by samwilson** [comment]
Sorry it's a nightmare. I'm trying to help!

Do you have `APP_ENV=dev` in `.env`? Does your webserver user have write access to the project's `var/` directory?

**Comment by WolfgangFahl** [comment]
What is APP_ENV=dev good for - which part needs that and why? 

```
/var/www/html$ ls -l
total 28
lrwxrwxrwx  1 wf wf          37 Feb 11 05:57 diagrams -> /home/wf/source/php/diagrams-service/

source/php/diagrams-service$ sudo chgrp -R www-data *
ls -l
total 404
drwxrwxr-x   2 wf www-data   4096 Feb 10 14:16 bin
-rw-rw-r--   1 wf www-data   2374 Feb 10 14:16 composer.json
-rw-rw-r--   1 wf www-data 155693 Feb 10 14:16 composer.lock
drwxrwxr-x   4 wf www-data   4096 Feb 10 14:16 config
-rw-rw-r--   1 wf www-data    609 Feb 10 14:16 Gruntfile.js
drwxrwxr-x   2 wf www-data   4096 Feb 10 14:16 i18n
drwxrwxr-x 387 wf www-data  16384 Feb 13 09:45 node_modules
-rw-rw-r--   1 wf www-data    342 Feb 13 09:43 package.json
-rw-rw-r--   1 wf www-data 164890 Feb 10 14:16 package-lock.json
-rw-rw-r--   1 wf www-data   1010 Feb 10 14:16 phpunit.xml.dist
drwxrwxr-x   2 wf www-data   4096 Feb 11 05:55 public
-rw-rw-r--   1 wf www-data   1669 Feb 10 14:16 README.md
drwxrwxr-x   3 wf www-data   4096 Feb 10 14:16 src
-rw-rw-r--   1 wf www-data   8691 Feb 10 14:16 symfony.lock
drwxrwxr-x   3 wf www-data   4096 Feb 10 14:16 templates
drwxrwxr-x   2 wf www-data   4096 Feb 10 14:16 tests
drwxrwxr-x   4 wf www-data   4096 Feb 10 14:20 var
drwxrwxr-x  15 wf www-data   4096 Feb 13 09:49 vendor
```

500 error persists ...

The apache access log has some 404:
```
2.0.0.25 - - [13/Feb/2020:09:49:32 +0100] "GET /diagrams/public/ HTTP/1.1" 200 2925 "http://fur.bitplan.com/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:72.0) Gecko/20100101 Firefox/72.0"
2.0.0.25 - - [13/Feb/2020:09:49:32 +0100] "GET /diagrams/public/style.css HTTP/1.1" 200 642 "http://fur.bitplan.com/diagrams/public/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:72.0) Gecko/20100101 Firefox/72.0"
2.0.0.25 - - [13/Feb/2020:09:49:32 +0100] "GET /diagrams/public/script.js HTTP/1.1" 200 1082 "http://fur.bitplan.com/diagrams/public/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:72.0) Gecko/20100101 Firefox/72.0"
127.0.0.1 - - [13/Feb/2020:09:49:36 +0100] "POST /diagrams/public/render HTTP/1.1" 404 438 "-" "Symfony HttpClient/Curl"
2.0.0.25 - - [13/Feb/2020:09:49:36 +0100] "POST /diagrams/public/ HTTP/1.1" 500 1018 "http://fur.bitplan.com/diagrams/public/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:72.0) Gecko/20100101 Firefox/72.0"
```

**Comment by WolfgangFahl** [comment]
There is no .env on my machine. Is that a directory or a file? I assume I need $HOME/env with a line APP_ENV=dev
```
cat .env
APP_ENV=dev
wf@fur:~$ sudo service apache2 restart
```
500 error persists

**Comment by WolfgangFahl** [comment]
https://symfony.com/what-is-symfony

« Symfony is a set of PHP Components, a Web Application framework, a Philosophy, and a Community — all working together in harmony. » 

👎 
I am now reading:
https://symfony.com/doc/current/deployment.html

**Comment by WolfgangFahl** [comment]
```
curl -sS https://get.symfony.com/cli/installer | bash
Symfony CLI installer

Environment check
  [*] cURL is installed
  [*] Gzip is installed
  [*] Git is installed

Download
  Finding the latest version (platform: "linux_amd64")...
  Downloading version 4.12.8 (https://github.com/symfony/cli/releases/download/v4.12.8/symfony_linux_amd64.gz)...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   613  100   613    0     0   2158      0 --:--:-- --:--:-- --:--:--  2150
100 7540k  100 7540k    0     0  3364k      0  0:00:02  0:00:02 --:--:-- 6185k
  Uncompress binary...
  Making the binary executable...
  Installing the binary into your home directory...
  The binary was saved to: /home/wf/.symfony/bin/symfony

The Symfony CLI v4.12.8 was installed successfully!

Add this to your shell configuration file:
  export PATH="$HOME/.symfony/bin:$PATH"
Start a new shell, and then run 'symfony'

Or install it globally on your system:
  mv /home/wf/.symfony/bin/symfony /usr/local/bin/symfony
and then run 'symfony'
wf@fur:~$ mv /home/wf/.symfony/bin/symfony /usr/local/bin/symfony
mv: cannot move '/home/wf/.symfony/bin/symfony' to '/usr/local/bin/symfony': Permission denied
wf@fur:~$ sudo mv /home/wf/.symfony/bin/symfony /usr/local/bin/symfony
wf@fur:~$ symfony check:requirements

Symfony Requirements Checker
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> PHP is using the following php.ini file:
/etc/php/7.2/cli/php.ini

> Checking Symfony requirements:

................................

                                              
 [OK]                                         
 Your system is ready to run Symfony projects 
```

**Comment by WolfgangFahl** [comment]
`
symfony server:start
`
let' s me use the diagram service on port 8000

When trying to use it it with an example it immediately hangs and gives
`PHP Fatal error:  Uncaught UnexpectedValueException: The stream or file "/home/wf/source/php/diagrams-service/var/log/prod.log" could not be opened: failed to open stream: Permission denied in /home/wf/source/php/diagrams-service/vendor/monolog/monolog/src/Monolog/Handler/StreamHandler.php:110 path="/usr/bin/php7.2" php="7.2.24"
`

which seems illogic given that the service runs with my user and the permissions are:
```
ls -l
total 4
-rw-rw-r-- 1 www-data www-data 3561 Feb 13 10:04 prod.log
```

in the log there seem to be some hints on the previous problems:
[`2020-02-13T10:04:37.978943+01:00] request.CRITICAL: Uncaught PHP Exception Symfony\Component\HttpClient\Exception\ClientException: "HTTP/1.1 404 Not Found returned for "http://fur.bitplan.com/diagrams/public/render"." at /home/wf/source/php/diagrams-service/vendor/symfony/http-client/Response/ResponseTrait.php line 261 {"exception":"[object] (Symfony\\Component\\HttpClient\\Exception\\ClientException(code: 404): HTTP/1.1 404 Not Found returned for \"http://fur.bitplan.com/diagrams/public/render\"`

**Comment by WolfgangFahl** [comment]
If i don't use the symbolic link but move the source directory the problem persists ...

**Comment by WolfgangFahl** [comment]
Now i found 
```
wf@fur:~/source/php/diagrams-service$ cat .env
APP_ENV=prod
APP_SECRET=secret-key-goes-here
PLANTUML_PATH=plantuml.jar
DIAGRAMS_DIR=var/diagrams/

```
and e.g. the plantuml.jar is obviously not shipped with the app - the var/diagrams folder is not existent so i created it - what is the APP_SECRET good for?

**Comment by WolfgangFahl** [comment]
After changing to "APP_ENV=dev" i now get error messages from Symfony.It's the 404 we already saw in the access log:
HTTP/1.1 404 Not Found returned for "http://fur.bitplan.com/diagrams/public/render".

**Comment by WolfgangFahl** [comment]
See http://diagrams.bitplan.com for an alternative. I'd like to make the API compatible - could you please comment on that in the gitter chat ...

**Comment by samwilson** [comment]
A compatible API sounds like a great idea!

For chat, I've been using [#mwstake-extensions:matrix.org](https://matrix.to/#/#mwstake-extensions:matrix.org); I'd rather not split out into yet another chat platform if I can avoid it. :-)

>  HTTP/1.1 404 Not Found returned for "http://fur.bitplan.com/diagrams/public/render".

I think this is likely to do with setting up URL rewriting. What web server are you using? There's an Apache .htaccess file in the `public/` directory.

**Comment by WolfgangFahl** [comment]
I have now updated https://www.mediawiki.org/wiki/Extension:Diagrams to show the two service options. I am not going to bother to get the Symfony stuff running any more. I am fine with the python solution which also has other benefits.  I have no clue what to do with https://matrix.to/#/#mwstake-extensions:matrix.org - it does not lead to a chat room.
I have added some issues to https://github.com/samwilson/diagrams-extension/issues which are more pressing for me right now since e.g. the image handling is a show stopper for me at this point. It would be good to get some more examples for http://wiki.bitplan.com/index.php/Diagrams#MediaWiki_Diagrams_Extension_Json_API
